File: Factory/Factory.ts
interface IVehicle {
  drive(): void;
}

class Car implements IVehicle {
  public drive() {
    console.log("Dirigindo um carro");
  }
}

class Bicycle implements IVehicle {
  public drive() {
    console.log("Pedalando uma bicicleta");
  }
}

class VehicleFactory {
  public static createVehicle(type: string): IVehicle {
    switch (type.toLowerCase()) {
      case 'car':
        return new Car();
      case 'bicycle':
        return new Bicycle();
      default:
        throw new Error('Tipo de veículo não suportado');
    }
  }
}

File: Factory/Factory.cs
namespace Factory;

interface IVehicle
{
  void Drive();
}

class Car : IVehicle
{
  public void Drive() => Console.WriteLine("Dirigindo um carro");
}

class Bicycle : IVehicle
{
  public void Drive() => Console.WriteLine("Pedalando uma bicicleta");
}

class VehicleFactory
{
  public static IVehicle CreateVehicle(string type) =>
    type.ToLower() switch
    {
      "car" => new Car(),
      "bicycle" => new Bicycle(),
      _ => throw new ArgumentException("Tipo de veículo não suportado")
    };
}
